tmc-langs-rust/src/main/java/fi/helsinki/cs/tmc/langs/rust/CargoPlugin.java

```diff
@@ -0,0 +1,137 @@
+import fi.helsinki.cs.tmc.langs.utils.ProcessRunner;
+
+import com.google.common.base.Optional;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Arrays;
+
+public class CargoPlugin extends AbstractLanguagePlugin {
+
+    private static final Logger log = LoggerFactory.getLogger(CargoPlugin.class);
+    private static final RunResult EMPTY_FAILURE = new RunResult(
+            Status.COMPILE_FAILED,
+            ImmutableList.<TestResult>of(),
+            new ImmutableMap.Builder<String, byte[]>().build());
+
+    public CargoPlugin() {
+        super(
+                new ExerciseBuilder(),
+                new StudentFileAwareSubmissionProcessor(),
+                new StudentFileAwareZipper(),
+                new StudentFileAwareUnzipper());
+    }
+
+    @Override
+    public boolean isExerciseTypeCorrect(Path path) {
+        return Files.isRegularFile(path.resolve(Constants.CARGO_TOML));
+    }
+
+    @Override
+    protected StudentFilePolicy getStudentFilePolicy(Path projectPath) {
+        return new CargoStudentFilePolicy(projectPath);
+    }
+
+    @Override
+    public ValidationResult checkCodeStyle(Path path) {
+        throw new UnsupportedOperationException("Not supported yet.");
+    }
+
+    @Override
+    public String getLanguageName() {
+        return "cargo";
+    }
+
+    public String getPluginName() {
+        return "cargo";
+    }
+
+    @Override
+    public Optional<ExerciseDesc> scanExercise(Path path, String exerciseName) {
+        throw new UnsupportedOperationException("Not supported yet.");
+    }
+
+    @Override
+    public RunResult runTests(Path path) {
+        Optional<RunResult> result = build(path);
+        if (result.isPresent()) {
+            log.info("Failed to compile project.");
+            return result.get();
+        }
+        return runBuiltTests(path);
+    }
+
+    private Optional<RunResult> build(Path dir) {
+        String[] command = {"cargo", "test", "--no-run"};
+        log.info("Building project with command {0}", Arrays.deepToString(command));
+        Optional<ProcessResult> result = run(command, dir);
+        if (result.isPresent()) {
+            if (result.get().statusCode == 0) {
+                return Optional.absent();
+            }
+            return Optional.of(filledFailure(result.get()));
+        }
+        return Optional.of(EMPTY_FAILURE);
+    }
+
+    private RunResult runBuiltTests(Path dir) {
+        String[] command = {"cargo", "test"};
+        log.info("Running tests with command {0}", Arrays.deepToString(command));
+        Optional<ProcessResult> result = run(command, dir);
+        if (result.isPresent()) {
+            return parseResult(result.get(), dir);
+        }
+        return EMPTY_FAILURE;
```

Comment:
`cargo test` failing here would mean that there was an IOException while running the command. I am not sure what situation this would happen that wouldn't already happen while building tests. And compilation output when compilation failure happens is already preserved.